---
title: "epigenetics"
author: "Aubrey Roberts"
date: "2023-02-26"
output: html_document
---

```{r setup, include=FALSE}
# Load packages
library(tidyverse)
library(ggplot2)
library(seqinr)
library(DESeq2)
library(EnhancedVolcano)
library(edgeR)
library(fgsea)
library(data.table)

# Access MoTrPAC data from this package
library(MotrpacRatTraining6moData)
library(MotrpacRatTraining6mo)
```

# Notes
```{r dict}
# We will look at chromatin accessibility (ATAC) and DNA methylation (METHYL). 

# DICTIONARY:

# Available data: 8 different tissues have ATAC data:
# HIPPOC: hippocampus
# BAT: brown adipose tissue
# HEART: heart
# LUNG: lung
# LIVER: liver
# WAT-SC: subcutaneous white adipose tissue
# KIDNEY: kidney
# SKM_GN: gastrocnemius (leg skeletal muscle)

# ATAC: chromatin accessibility
# METHYL: DNA methylation, RRBS data


# ANALYSIS PLAN:

# 1) Do PCA plot with response to exercise as the focus ("group"); make it color gradient not a factor -- pick tissue with biggest exercise separation gradient (control vs. 1 vs 2. vs. 4 vs. 8 weeks). We chose kidney tissue.

# 2) Use table browser from UCSC. Download promoter regions from UCSC 1kb region upstream. Look at which of our peaks (row names) correspond to peaks near promoters. These will likely be most interesting/relevant to gene expression. Use bedtools (ask prof to share slides for bed tools).

# 3) Use two methods for differential peak analysis. Feed it the peaks by promoter regions we identified. 
# a) With normalized data: wilcoxon rank sum and benj hochberg correction
# b) With unnormalized data: DESeq2 -- does multiple hypothesis testing on everything 

# 4) Look at peaks differentially open in promoter regions and do pathway analysis. Compare wilcoxon rank sum and DESeq2 output.

```

# Load Kidney Data
```{r kidneydata}

setwd("/Users/aubreyroberts/Documents/GitHub/biods205/")

# Download and load in epigenetic data from MoTrPAC package website: https://github.com/MoTrPAC/MotrpacRatTraining6moData#about-this-package

kidney<-load(file="ATAC_KIDNEY_NORM_DATA.rda")
kidney_raw<-load(file="ATAC_KIDNEY_RAW_COUNTS.rda")
annotation<-load(file="ATAC_FEATURE_ANNOT.rda")

# We now have loaded in two new frames of counts:

# Normalized - ATAC_KIDNEY_NORM_DATA

head(ATAC_KIDNEY_NORM_DATA)

# Raw - ATAC_KIDNEY_RAW_COUNTS

head(ATAC_KIDNEY_RAW_COUNTS)

```

# PCA plot for kidney - MAKE INTO COLOR GRADIENT
```{r kidneyPCA}
         
# Create PCA plot of normalized data

# First four columns are metadata, so we don't need in normalized count matrix
sample_ids = colnames(ATAC_KIDNEY_NORM_DATA)[-c(1:4)]

# Perform PCA
pca = stats::prcomp(t(ATAC_KIDNEY_NORM_DATA[,sample_ids]))
  
# Make a data frame with some phenotypic data and the first 3 PCs
df = data.frame(
  group = PHENO[sample_ids,"group"],
  sex = PHENO[sample_ids,"sex"],
  pca$x[,1:3] # take the first principal components
)

# Plot the first two PCs
print(ggplot(df, aes(x=PC1, y=PC2, color=group, shape=sex)) +
  geom_point(size=3))+ggtitle("Kidney")+theme(plot.title = element_text(hjust = 0.5))

```
# Select for peaks of interest (near promoter regions)
```{r}

# Filter for peaks of interest near promoters:

# Ideally, we will filter for peaks in promoter region before this -- then do differential expression below.
# Use "upstream1000.fa.gz" file from UCSC genome browser: https://hgdownload.soe.ucsc.edu/goldenPath/rn7/bigZips/ to filter for peaks of interest in promoter regions
# Need to convert this file into a bed file, use bedtools to see which of our peaks fall in given regions in file.
# Then feed these peaks of interest into DESeq2.

# Load annotated peak data
annotation_concise <- ATAC_FEATURE_ANNOT %>% select("feature_ID","chrom","start","end")

dim(ATAC_FEATURE_ANNOT)
dim(ATAC_KIDNEY_RAW_COUNTS)

# Join our data frame of our peaks with annotated data
combined_data <- left_join(annotation_concise, ATAC_KIDNEY_RAW_COUNTS) %>% select(-c("feature","assay","tissue"))

vec<-unique(annotation_concise$feature_ID)
all(vec %in% ATAC_KIDNEY_RAW_COUNTS$feature_ID)
all(ATAC_KIDNEY_RAW_COUNTS$feature_ID %in% vec)

all(annotation_concise$feature_ID)
head(ATAC_KIDNEY_RAW_COUNTS)
head(annotation_concise)

```

# Prepare bed files
```{r}
bed_data <- annotation_concise %>% select(-c("feature_ID"))
bed_data <- bed_data %>% mutate(chrom=paste("chr",chrom,sep = ""))
bed_data$start <- as.integer(bed_data$start)
bed_data$end <- as.integer(bed_data$end)
write.table(bed_data, "kidney_peaks.bed", sep="\t", col.names=FALSE, row.names = FALSE, append = TRUE, quote = FALSE) 

promoter_data <- read.table("fake_bed.csv",sep=',')[-1,]
promoter_data$V2 <- as.integer(promoter_data$V2)
promoter_data$V3 <- as.integer(promoter_data$V3)
write.table(promoter_data, "promoter.bed", sep="\t", col.names=FALSE, row.names = FALSE, append = TRUE, quote = FALSE) 
```

# Intersect peaks with promoter region
```{r}
combined_data <- combined_data %>% mutate(chrom=paste("chr",chrom,sep = ""))

# bedtools intersect -a kidney_peaks.bed -b promoter.bed -wa > intersect.bed

promoter_peaks <- read.table("intersect.bed",sep="\t")
colnames(promoter_peaks) <- c("chrom","start","end")

filtered_peaks <- left_join(promoter_peaks, combined_data)

```

# Keep peaks that overlap with promoter region
```{r}
# Filter for peaks of interest

filtered_peaks_dist <- filtered_peaks %>% distinct()

# Set features to rownames to format filtered raw counts for DESeq2

rownames(filtered_peaks_dist) <- filtered_peaks_dist$feature_ID

filtered_raw_counts <- filtered_peaks_dist[,-c(1:4)]

dim(filtered_raw_counts)

```

# Run DESeq2 for differential expression of promoter peaks
```{r deseq}
# DIFFERENTIALLY EXPRESSED PEAKS WITH DESEQ2
# Filter for phenodata for kidney samples in filtered raw counts
sampleIDs<-colnames(filtered_raw_counts)

kidney_PHENO<-PHENO %>% 
  filter(rownames(PHENO) %in% sampleIDs)

# Make sure order of samples is the same
kidney_PHENO<-arrange(kidney_PHENO, viallabel)

rownames(kidney_PHENO) == colnames(filtered_raw_counts)

head(filtered_raw_counts)
dim(kidney_PHENO)

# Run DESeq2 on filtered raw counts and phenodata
dds <- DESeqDataSetFromMatrix(countData = filtered_raw_counts, 
                              colData = kidney_PHENO,
                              design = ~group)

dds <- DESeq(dds)

res <- results(dds)

summary(res)

plotMA(res)

# Look at contrasts between groups
unique(PHENO$group)

# Control vs. 1 wk: results and top 10 by adjusted p-value

control_1w <-results(dds, contrast=c("group","control","1w")) 
summary(control_1w)
head(control_1w[order(control_1w$padj),], 10)


# Control vs. 2 week
control_2w <- results(dds, contrast=c("group", "control", "2w"))
head(control_2w[order(control_2w$padj),], 10)


# Control vs. 8 week
control_8w <- results(dds, contrast=c("group", "control", "8w"))
summary(control_8w)
head(control_8w[order(control_8w$padj),], 10)


#Volcano for Control vs. 8 week
control_8w <- control_8w[order(res$padj),]
rownames(control_8w) <- control_8w$row
head(control_8w)

EnhancedVolcano(control_8w,
    lab = rownames(control_8w),
    x = 'log2FoldChange',
    y = 'padj')

```


# Moving forward...
```{r}
## INSTRUCTIONS FOR HOW TO PROCEED:

# 1) INTERSECT the FDR normalized peaks with the promoter region to see if those 237 regions are in promoter --- DONE.
# 237 DEGs from main paper
any(filtered_raw_counts[ATAC_KIDNEY_NORM_DATA_05FDR$feature_ID,])
# 6 DEGs they found are in promoter region -- they used limma on all features?
sum(ATAC_KIDNEY_NORM_DATA_05FDR$feature_ID %in% rownames(filtered_raw_counts))

# 1) troubleshoot PREPROCESS for DESeq2 - need to do some normalization; size factor adjustment/normalization

# 2) DO WILCOXON FOLLOWED BY BENJ HOCH -- DONE.

# 3) limma or edgeR for bulk atac seq as well

# 4) compare the techniques

# 5) bed merge - can look for overlapping segments in our kidney data; sum the peaks and exclude overlapping peaks; tool in bedtools to sum to get single value for gene --- DONE.


## 3/7/23 Notes:
# try wilcoxin rank sum on promoter region AND all peaks (even non-promoter)?
# prerank GSEA -- log fold change for unique genes (collapse all peaks into unique genes first); see if any gene sets are more skewed to one phenotype/timepoint
# fgsea package?
# Look up how they did limma for MoTrPAC and got the DE peaks
```
# LIMMA
```{r limma}

# Create design matrix with model.matrix() -- specifies which variables/coefficients will be in linear model
design <- model.matrix(~0+kidney_PHENO$group+kidney_PHENO$sex)
colnames(design) <-c("oneweek", "twoweek", "fourweek", "eightweek", "control", "male")
design

# Specify contrasts with makeContrasts()
contr.matrix <- makeContrasts(
   control_vs_eight = control - eightweek,
   control_vs_four = control - fourweek,
   control_vs_two = control - twoweek,
   control_vs_one = control - oneweek,
   levels = colnames(design))
contr.matrix

#In limma, linear modelling is carried out on the log-CPM values which are assumed to be normally distributed and the mean-variance relationship is accommodated using precision weights calculated by the voom function.
v<-voom(filtered_raw_counts, design=design, plot=TRUE)
vfit <- lmFit(v, design)
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)
efit <- eBayes(vfit)
summary(decideTests(efit))

# View top DEGs with topTreat()
topTreat(efit)

# Highest log fold changes:
top.table<-topTreat(efit, sort.by = "logFC", n=Inf)
head(top.table)
length(which(top.table$adj.P.Val < 0.05)) # number of DEGs

```

# NORMALIZATION
# Normalize filtered_raw_counts to "filtered_norm_counts" for Wilcoxin rank sum (adapted from Bootcamp)
```{r}
# Remove genes with 0 counts in ALL samples
# sum gene expression across samples
genesum <- rowSums(filtered_raw_counts)

# subset dataframe
filtered_norm_counts <- filtered_raw_counts[genesum>0,] # filters out 2 peaks
dim(filtered_raw_counts)
dim(filtered_norm_counts)

# CPM normalization
# sum all counts per sample
samplesums <- colSums(filtered_norm_counts)

# divide sum of each sample (column) by 1,000,000 to derive scalefactor
scalefactor <- samplesums/1000000

# divide expression in each sample (column) by corresponding scalefactor
filtered_norm_counts <- t(t(filtered_norm_counts) / scalefactor)  #transposition is important to divide across columns

#check that all samples summed expression == 1M
colSums(filtered_norm_counts)

# Log2 Normalization
# use +1 because can't take logs of zero
filtered_norm_counts <- log2(filtered_norm_counts + 1)
head(filtered_norm_counts)

```

# WILCOXON RANK SUM
# Differential Expression with two-sided Wilcoxon rank sums test followed by Benjamini-Hochberg adjustment of p-values
```{r wilcox}
# Want to use normalized data for this.

# we want to compare pairs of phenotypes: control, 1wk, 2wk, 4wk, 8wk --- four pairwise contrasts (control v 1, control v 2, control v 4, control v 8)

# create subsets of IDs 
controlids <- rownames(kidney_PHENO[kidney_PHENO$group=="control",])
week1_ids <- rownames(kidney_PHENO[kidney_PHENO$group=="1w",])
week2_ids <- rownames(kidney_PHENO[kidney_PHENO$group=="2w",])
week4_ids <- rownames(kidney_PHENO[kidney_PHENO$group=="4w",])
week8_ids <- rownames(kidney_PHENO[kidney_PHENO$group=="8w",])

# Note: filtered_norm_counts stores normalized count data for promoter regions only; could also try this with ATAC_KIDNEY_NORM_DATA (normalized data for all peaks, even outside promoter)

# Run Wilcoxon rank sum test (control vs. 4 wk):

todf <- list()
# for loop through all genes
allgenes <- rownames(ATAC_KIDNEY_NORM_DATA)
for (i in seq(length(allgenes))){
  
  # calculate log2fc between control and 4 wk
  gene <- allgenes[i]
  control <- as.numeric(ATAC_KIDNEY_NORM_DATA[i,controlids])
  wk4 <- as.numeric(ATAC_KIDNEY_NORM_DATA[i,week4_ids])
  log2fc <- mean(control) - mean(wk4)   #log(control/wk4) = log(control)-log(4wk)
  
  # wilcoxon rank sums - outputs tstatistic and pvalue; extract pvalue
  res <- wilcox.test(control, wk4, alternative = "two.sided", exact=FALSE)
  pvalue <- res$p.value
  
  # store results
  todf[[i]] <- c(gene, log2fc, pvalue)
}

# reformat list of vectors into dataframe
dfstats <- as.data.frame(do.call(rbind, todf))
colnames(dfstats) <- c("gene", "log2fc", "pvalue")
dfstats$log2fc <- as.numeric(dfstats$log2fc)
dfstats$pvalue <- as.numeric(dfstats$pvalue)
dfstats <- dfstats[order(dfstats$pvalue),]
head(dfstats)

# multiple hypothesis correction using benjamini-hochberg procedure
dfstats$qvalue <- p.adjust(dfstats$pvalue, method="BH")
dfstats[(dfstats$qvalue<0.05),]
any(dfstats$qvalue < 0.05) # NO SIGNIFICANT P-VALUES AFTER ADJUSTMENT 

# volcano plot with p value
EnhancedVolcano(dfstats,
    lab = dfstats$gene,
    x = 'log2fc',
    y = 'qvalue',
    title = 'Differential Expression Control/4wk')
```

# GSEA
# Fast preranked gene set enrichment analysis (GSEA) with LFC of genes
```{r gsea}

library(fgsea)
# Good resource: https://bioinformatics-core-shared-training.github.io/RNAseq_September_2019/html/06_Gene_set_testing.html#fgsea

# Step 1: Ranking Data. Provide fgsea a vector containing numeric data by which it should rank the genes. 
# We can rank based on fold change.

rankData<-top.table$logFC
names(rankData) <- rownames(top.table)

# Step 2: Load in pathways (external source?)

# Step 3: Conduct analysis
fgseaRes <- fgsea(pathways,
                  rankData,
                  minSize = 15,
                  maxSize = 500,
                  nperm = 1000)

# use plotEnrichment() later to plot enrichment scores
# use plotGseaTable() to plot a summary figure showing pathways


```






```






