---
title: "Epigenetics -  Kidney"
author: "Aubrey Roberts"
date: "2023-02-26"
output: html_document
---

```{r setup, include=FALSE}
# Load packages
library(tidyverse)
library(ggplot2)
library(seqinr)
library(DESeq2)
library(EnhancedVolcano)
library(edgeR)
library(fgsea)
library(data.table)

# Access MoTrPAC data from this package
library(MotrpacRatTraining6moData)
library(MotrpacRatTraining6mo)
```

# Background:

# We will look at chromatin accessibility (ATAC) and DNA methylation (METHYL). 

# Available data: 8 different tissues have ATAC data:
HIPPOC: hippocampus
BAT: brown adipose tissue
HEART: heart
LUNG: lung
LIVER: liver
WAT-SC: subcutaneous white adipose tissue
KIDNEY: kidney
SKM_GN: gastrocnemius (leg skeletal muscle)

# ATAC: chromatin accessibility
# METHYL: DNA methylation, RRBS data


# ANALYSIS PLAN:

1) Do *PCA plot* with response to exercise as the focus ("group"); make it color gradient not a factor -- pick tissue with biggest exercise separation gradient (control vs. 1 vs 2. vs. 4 vs. 8 weeks). We chose kidney tissue.

2) *Filter for peaks in promoter region.* Use table browser from UCSC. Download promoter regions from UCSC 1kb region upstream. Look at which of our peaks (row names) correspond to peaks near promoters. Use bedtools. These will likely be most interesting/relevant to gene expression. 

3) Try out different methods to look for *Differentially Expressed Peaks.* Feed it the peaks by promoter regions we identified. 
          a) *DESeq2*: use unnormalized data--does multiple hypothesis testing on everything 
          b) *Limma* / voom
          c) *Wilcoxon rank sum* and benj hochberg correction: use normalized data
Compare DESeq2, limma, and wilcoxon rank sum output.

4) Look at peaks differentially open in promoter regions and do *pathway analysis (GSEA)* based on LFC. 


# Load Kidney Data
Download and load in epigenetic data from MoTrPAC package website: https://github.com/MoTrPAC/MotrpacRatTraining6moData#about-this-package
```{r kidneydata}

setwd("/Users/aubreyroberts/Documents/GitHub/biods205/")

kidney<-load(file="data/ATAC_KIDNEY_NORM_DATA.rda")
kidney_raw<-load(file="data/ATAC_KIDNEY_RAW_COUNTS.rda")
annotation<-load(file="data/ATAC_FEATURE_ANNOT.rda")
#dea_results<-load(file="ATAC_KIDNEY_DA.rda")

# We now have loaded in two new frames of counts:

# Normalized - ATAC_KIDNEY_NORM_DATA
head(ATAC_KIDNEY_NORM_DATA)

# Raw - ATAC_KIDNEY_RAW_COUNTS
head(ATAC_KIDNEY_RAW_COUNTS)

```

# PCA plot for kidney 
```{r kidneyPCA}

# Create PCA plot of normalized data

# First four columns are metadata, so we don't need in normalized count matrix
sample_ids = colnames(ATAC_KIDNEY_NORM_DATA)[-c(1:4)]

# Perform PCA
pca = stats::prcomp(t(ATAC_KIDNEY_NORM_DATA[,sample_ids]))
  
# Make a data frame with some phenotypic data and the first 3 PCs
df = data.frame(
  group = PHENO[sample_ids,"group"],
  sex = PHENO[sample_ids,"sex"],
  pca$x[,1:3] # take the first principal components
)

# Plot the first two PCs
print(ggplot(df, aes(x=PC1, y=PC2, color=group, shape=sex)) +
  geom_point(size=3))+
  ggtitle("Kidney")+
  theme(plot.title = element_text(hjust = 0.5))+
  scale_color_manual(values=GROUP_COLORS)

```

# Filter for peaks near promoter regions. 
Use UCSC genome browser file and bedtools. Filter dataset for promoter peaks.
```{r}
# Filter for peaks of interest near promoters:

# Use "upstream1000.fa.gz" file from UCSC genome browser: https://hgdownload.soe.ucsc.edu/goldenPath/rn7/bigZips/ to filter for peaks of interest in promoter regions

# Need to convert this file into a bed file, use bedtools to see which of our peaks fall in given regions in file.

# Then feed these peaks of interest into DESeq2.

# Load annotated peak data
annotation_concise <- ATAC_FEATURE_ANNOT[,c("feature_ID","chrom","start","end")]

dim(ATAC_FEATURE_ANNOT)
dim(ATAC_KIDNEY_RAW_COUNTS)

# Join our data frame of our peaks with annotated data
combined_data <- left_join(annotation_concise, ATAC_KIDNEY_RAW_COUNTS) %>% dplyr::select(-c("feature","assay","tissue"))
combined_data <- combined_data %>% mutate(chrom=paste("chr",chrom,sep = ""))

vec<-unique(annotation_concise$feature_ID)
all(vec %in% ATAC_KIDNEY_RAW_COUNTS$feature_ID)
all(ATAC_KIDNEY_RAW_COUNTS$feature_ID %in% vec)

all(annotation_concise$feature_ID)
head(ATAC_KIDNEY_RAW_COUNTS)
head(annotation_concise)

# Prepare bed files: the code below is how we created the bed files, it is now commented out as now we just need to read in our final files that are created rather than remake them each time

# bed_data <- annotation_concise %>% select(-c("feature_ID"))
# bed_data <- bed_data %>% mutate(chrom=paste("chr",chrom,sep = ""))
# bed_data$start <- as.integer(bed_data$start)
# bed_data$end <- as.integer(bed_data$end)
#write.table(bed_data, "kidney_peaks_new.bed", sep="\t", col.names=FALSE, row.names = FALSE, append = TRUE, quote = FALSE) 

# fake_bed.csv is created by parsing the UCSC genome browser FASTA file for sequences within 1000 bp of promoters to extract the chromosome, start, and end location 

# promoter_data <- read.table("fake_bed.csv",sep=',')[-1,]
# promoter_data$V2 <- as.integer(promoter_data$V2)
# promoter_data$V3 <- as.integer(promoter_data$V3)
# write.table(promoter_data, "promoter.bed", sep="\t", col.names=FALSE, row.names = FALSE, append = TRUE, quote = FALSE) 

# Intersect peaks with promoter region
# Line for intersection: bedtools intersect -a kidney_peaks.bed -b promoter.bed -wa > intersect.bed
# promoter_peaks <- read.table("intersect.bed",sep="\t")
# colnames(promoter_peaks) <- c("chrom","start","end")

# We also checked if there is any overlap in our peaks
# combined_data <- combined_data %>% mutate(chrom=paste("chr",chrom,sep = ""))
# sortbed -i kidney_peaks_new.bed > kidney_peaks_sorted_new.bed
# bedtools merge -i kidney_peaks_sorted_new.bed > kidney_overlap_check_new.bed
# Merge shows no overlaps! Yay!


# Keep peaks that overlap with promoter region/filter for peaks of interest:
promoter_peaks <- read.table("atac_promoter_peaks.bed")
colnames(promoter_peaks) <- c("chrom","start","end")
filtered_peaks <- left_join(promoter_peaks, combined_data)
filtered_peaks_dist <- filtered_peaks %>% distinct()

# Set features to rownames to format filtered raw counts for DESeq2
rownames(filtered_peaks_dist) <- filtered_peaks_dist$feature_ID
filtered_raw_counts <- filtered_peaks_dist[,-c(1:4)]
head(filtered_raw_counts)

```

# Prep for DESeq2
Make sure phenotype and count data line up. Subset data for separate analyses of male vs. female.
```{r deseq}

# Filter for phenodata for kidney samples in filtered raw counts
sampleIDs<-colnames(filtered_raw_counts)

kidney_PHENO<-PHENO %>% 
  filter(rownames(PHENO) %in% sampleIDs)

# Make sure order of samples is the same
kidney_PHENO<-arrange(kidney_PHENO, viallabel)

#rownames(kidney_PHENO) == colnames(filtered_raw_counts)

# Add metadata covariates (includes Sample_batch, ...frip) to kidney_PHENO

rownames(ATAC_META)<-ATAC_META$viallabel
ATAC_META$viallabel <- NULL

sampleIDs<-rownames(kidney_PHENO)
kidney_META<-ATAC_META[rownames(ATAC_META) %in% sampleIDs,]

any(rownames(kidney_META) != rownames(kidney_PHENO))

kidney_PHENO2<-cbind(kidney_META, kidney_PHENO)
saverows<-rownames(kidney_PHENO2)

# Important: convert kidney to a factor and reorder levels
kidney_PHENO2$group <- factor(kidney_PHENO2$group, levels = c("control", "1w", "2w", "4w", "8w"))
unique(kidney_PHENO2$group)
kidney_PHENO2$sex<-as.factor(kidney_PHENO2$sex)
kidney_PHENO2$Sample_batch<-as.factor(kidney_PHENO2$Sample_batch)

```

# DESeq2
For differential expression of promoter peaks (Interaction term for sex*group)
Note: No significant results.
```{r de}

# Run DESeq2 on filtered raw counts and phenodata
dds <- DESeqDataSetFromMatrix(countData = filtered_raw_counts, 
                              colData = kidney_PHENO2,
                              design = ~Sample_batch+peak_enrich.frac_reads_in_peaks.macs2.frip+sex+sex*group+group)

dds <- DESeq(dds)
res <- results(dds)

# Control vs. 1 wk: results and top 10 by adjusted p-value

control_1w <-results(dds, contrast=c("group","control","1w")) 
summary(control_1w)
head(control_1w[order(control_1w$padj),], 10)

# Control vs. 2 week
control_2w <- results(dds, contrast=c("group", "control", "2w"))
summary(control_2w)
head(control_2w[order(control_2w$padj),], 10)

# Control vs. 4 week
control_4w <- results(dds, contrast=c("group", "control", "4w"))
summary(control_4w)
head(control_4w[order(control_4w$padj),], 10)

# Control vs. 8 week
control_8w <- results(dds, contrast=c("group", "control", "8w"))
summary(control_8w)
head(control_8w[order(control_8w$padj),], 10)


# VISUALIZATION OF P-VALUES:
# Make control_8w contrast results into dataframe
control_8w<-data.frame(control_8w)
control_8w$pbins<-NULL
control_8w$pbins[control_8w$padj > 0 & control_8w$padj < 0.05] <- "<0.05"
control_8w$pbins[control_8w$padj >=0.05 & control_8w$padj < 0.3] <- "0.05-0.3"
control_8w$pbins[control_8w$padj >=0.30 & control_8w$padj < 0.6] <- "0.3-0.6"
control_8w$pbins[control_8w$padj >=0.60 & control_8w$padj < 0.9] <- "0.6-.9"
control_8w$pbins[control_8w$padj >=0.90 & control_8w$padj < 1] <- "0.9-1.0"

control_8w<-control_8w %>% 
  filter(is.na(padj) == FALSE)

ggplot(data = control_8w, 
       aes(x = pbins)) + 
  geom_histogram(stat = "count", color="light gray", fill="light gray") +
  stat_count(binwidth = 1, 
             geom = 'text', 
             color = 'black', 
             aes(label = ..count..),
           position = position_stack(vjust = 0.5))+
  scale_fill_discrete(drop=FALSE) +
          scale_x_discrete(drop=FALSE)+
  ggtitle("DESeq2: Adjusted p-value distribution (Control vs. 8 wk)")+
  theme(plot.title = element_text(hjust = 0.5))+
  labs(x="Adjusted p-value bins", y="Number peaks")

#control_8w %>% 
  #filter(padj < 0.9)
```

# We didn't find much with DESeq2.
# Moving forward...things to try:

1a) INTERSECT the FDR normalized peaks with the promoter region to see if those 237 regions are in promoter --- DONE.
There are 237 DEGs from main paper. We found that only 6 of these were in promoter region.

any(filtered_raw_counts[ATAC_KIDNEY_NORM_DATA_05FDR$feature_ID,])
sum(ATAC_KIDNEY_NORM_DATA_05FDR$feature_ID %in% rownames(filtered_raw_counts))

1b) Look up how they did limma for MoTrPAC and got the DE peaks. Try to reproduce analysis and get at least 6 DE peaks with limma.

2) Troubleshoot / PREPROCESS for DESeq2 - need to do some normalization; size factor adjustment/normalization.

3) DO WILCOXON FOLLOWED BY BENJ HOCH -- DONE.

4) Try limma or edgeR as well.

5) Compare the techniques.

6) Bed merge - can look for overlapping segments in our kidney data; sum the peaks and exclude overlapping peaks; tool in bedtools to sum to get single value for gene --- DONE.
We found that there is not overlap which is good.

7) Use prerank GSEA -- log fold change for unique genes (collapse all peaks into unique genes first); see if any gene sets are more skewed to one phenotype/timepoint. (fgsea package).


# Limma - try to reproduce MoTrPAC findings.
Try to replicate original MoTrPAC findings with limma:
1) use normalize.method="quantile" in voom
2) make sex_group contrasts
3) use model with covariates (sample_batch and peak_enrich.frac_reads_in_peaks.macs2.frip)

```{r limma revamp}

# helpful: https://seqqc.wordpress.com/2020/11/28/10-tips-tricks-for-complex-model-matrix-designs-in-dge-analysis/

# Create design matrix; use 0 as intercept because interested in pairwise contrasts
kidney_PHENO2<-data.table(kidney_PHENO2)
kidney_PHENO2[,sex_group :=paste0(sex, "_", group)]
kidney_PHENO2[,sex_group :=factor(sex_group)]

full_model_str = "~ 0 + sex_group + Sample_batch + peak_enrich.frac_reads_in_peaks.macs2.frip"
design = model.matrix(as.formula(full_model_str), data=kidney_PHENO2)
design

# Limma model
v = voom(filtered_raw_counts, design, normalize.method="quantile")
limma_model1 = lmFit(v, design)

cont.matrix=makeContrasts(
  control_1wk_M = sex_groupmale_1w - sex_groupmale_control,
  control_2wk_M = sex_groupmale_2w - sex_groupmale_control,
  control_4wk_M = sex_groupmale_4w - sex_groupmale_control,
  control_8wk_M = sex_groupmale_8w - sex_groupmale_control,
  control_1wk_F = sex_groupfemale_1w - sex_groupfemale_control,
  control_2wk_F = sex_groupfemale_2w - sex_groupfemale_control,
  control_4wk_F = sex_groupfemale_4w - sex_groupfemale_control,
  control_8wk_F = sex_groupfemale_8w - sex_groupfemale_control,
  levels=design)
  
fit2=contrasts.fit(limma_model1,cont.matrix)
limma_res=eBayes(fit2)
topTable(limma_res, coef = "control_1wk_M")
topTable(limma_res, coef = "control_2wk_M")
topTable(limma_res, coef = "control_4wk_M")
male8wk<-topTable(limma_res, coef = "control_8wk_M", n=Inf)
topTable(limma_res, coef = "control_1wk_F")
topTable(limma_res, coef = "control_2wk_F")
topTable(limma_res, coef = "control_4wk_F")
female8wk<-topTable(limma_res, coef = "control_8wk_F", n=Inf)


# VISUALIZATION OF P-VALUES:
# Make control_8w contrast results into dataframe
female8wk
female8wk$pbins<-NULL
female8wk$pbins[female8wk$adj.P.Val > 0 & female8wk$adj.P.Val < 0.05] <- "<0.05"
female8wk$pbins[female8wk$adj.P.Val >=0.05 & female8wk$adj.P.Val < 0.3] <- "0.05-0.3"
female8wk$pbins[female8wk$adj.P.Val >=0.30 & female8wk$adj.P.Val < 0.6] <- "0.3-0.6"
female8wk$pbins[female8wk$adj.P.Val >=0.60 & female8wk$adj.P.Val < 0.9] <- "0.6-.9"
female8wk$pbins[female8wk$adj.P.Val >=0.90 & female8wk$adj.P.Val < 1] <- "0.9-1.0"

female8wk<-female8wk %>% 
  filter(is.na(adj.P.Val) == FALSE)

ggplot(data = female8wk, 
       aes(x = pbins)) + 
  geom_histogram(stat = "count", color="lightpink", fill="lightpink") +
  stat_count(binwidth = 1, 
             geom = 'text', 
             color = 'black', 
             aes(label = ..count..),
           position = position_stack(vjust = 0.5))+
  scale_fill_discrete(drop=FALSE) +
          scale_x_discrete(drop=FALSE)+
  ggtitle("Limma: Adjusted p-value distribution (female control vs. female 8 wk)")+
  theme(plot.title = element_text(hjust = 0.5))+
  labs(x="Adjusted p-value bins", y="Number peaks")


```

# Normalization prior to Wilcoxon Rank Sum
Normalize filtered_raw_counts to "filtered_norm_counts" for Wilcoxin rank sum (adapted from Bootcamp)
```{r}
# Remove genes with 0 counts in ALL samples
# sum gene expression across samples
genesum <- rowSums(filtered_raw_counts)

# subset dataframe
filtered_norm_counts <- filtered_raw_counts[genesum>0,] # filters out 2 peaks
dim(filtered_raw_counts)
dim(filtered_norm_counts)

# CPM normalization
# sum all counts per sample
samplesums <- colSums(filtered_norm_counts)

# divide sum of each sample (column) by 1,000,000 to derive scalefactor
scalefactor <- samplesums/1000000

# divide expression in each sample (column) by corresponding scalefactor
filtered_norm_counts <- t(t(filtered_norm_counts) / scalefactor)  #transposition is important to divide across columns

#check that all samples summed expression == 1M
#colSums(filtered_norm_counts)

# Log2 Normalization
# use +1 because can't take logs of zero
filtered_norm_counts <- log2(filtered_norm_counts + 1)

```

# Wilcoxon Rank Sum Test

Differential Expression with two-sided Wilcoxon rank sums test followed by Benjamini-Hochberg adjustment of p-values
```{r wilcox}
# Want to use normalized data for this.

# Subset phenodata and get female and male IDs
kidney_PHENO2<-as.data.frame(kidney_PHENO2)
rownames(kidney_PHENO2)<-saverows
female<-kidney_PHENO2 %>% 
  filter(sex == "female")
f_ids<-rownames(female)

male<-kidney_PHENO2 %>% 
  filter(sex == "male")
m_ids<-rownames(male)

# Subset counts to female and male
M_filtered_norm_counts <- filtered_norm_counts[,colnames(filtered_raw_counts) %in% m_ids]
F_filtered_norm_counts <- filtered_norm_counts[,colnames(filtered_raw_counts) %in% f_ids]

# we want to compare pairs of phenotypes: control, 1wk, 2wk, 4wk, 8wk --- four pairwise contrasts (control v 1, control v 2, control v 4, control v 8)

# MALE
# create subsets of IDs 
controlids <- rownames(male[male$group=="control",])
week1_ids <- rownames(male[male$group=="1w",])
week2_ids <- rownames(male[male$group=="2w",])
week4_ids <- rownames(male[male$group=="4w",])
week8_ids <- rownames(male[male$group=="8w",])

# Note: filtered_norm_counts stores normalized count data for promoter regions only; could also try this with ATAC_KIDNEY_NORM_DATA (normalized data for all peaks, even omoter)
# Run Wilcoxon rank sum test (control vs. 8 wk):

todf <- list()
# for loop through all genes
allgenes <- rownames(M_filtered_norm_counts)
for (i in seq(length(allgenes))){
  
  # calculate log2fc between control and 4 wk
  gene <- allgenes[i]
  control <- as.numeric(M_filtered_norm_counts[i,controlids])
  wk8 <- as.numeric(M_filtered_norm_counts[i,week8_ids])
  log2fc <- mean(control) - mean(wk8)   #log(control/wk8) = log(control)-log(4wk)
  
  # wilcoxon rank sums - outputs tstatistic and pvalue; extract pvalue
  res <- wilcox.test(control, wk8, alternative = "two.sided", exact=FALSE)
  pvalue <- res$p.value
  
  # store results
  todf[[i]] <- c(gene, log2fc, pvalue)
}

# reformat list of vectors into dataframe
dfstats <- as.data.frame(do.call(rbind, todf))
colnames(dfstats) <- c("gene", "log2fc", "pvalue")
dfstats$log2fc <- as.numeric(dfstats$log2fc)
dfstats$pvalue <- as.numeric(dfstats$pvalue)
dfstats <- dfstats[order(dfstats$pvalue),]

# multiple hypothesis correction using benjamini-hochberg procedure
dfstats$qvalue <- p.adjust(dfstats$pvalue, method="BH")
head(dfstats)
dfstats[(dfstats$qvalue<0.05),]
any(dfstats$qvalue < 0.05) # NO SIGNIFICANT P-VALUES AFTER ADJUSTMENT 


# VISUALIZATION OF P-VALUES:
# Make control_8w contrast results into dataframe
dfstats$pbins<-NULL
dfstats$pbins[dfstats$qvalue > 0 & dfstats$qvalue < 0.05] <- "<0.05"
dfstats$pbins[dfstats$qvalue >=0.05 & dfstats$qvalue < 0.3] <- "0.05-0.3"
dfstats$pbins[dfstats$qvalue >=0.30 & dfstats$qvalue < 0.6] <- "0.3-0.6"
dfstats$pbins[dfstats$qvalue >=0.60 & dfstats$qvalue < 0.9] <- "0.6-.9"
dfstats$pbins[dfstats$qvalue >=0.90 & dfstats$qvalue <= 1] <- "0.9-1.0"

dfstats<-dfstats %>% 
  filter(is.na(qvalue) == FALSE)

ggplot(data = dfstats, 
       aes(x = pbins)) + 
  geom_histogram(stat = "count", color="lightblue", fill="lightblue") +
  stat_count(binwidth = 1, 
             geom = 'text', 
             color = 'black', 
             aes(label = ..count..),
           position = position_stack(vjust = 0.5))+
  scale_fill_discrete(drop=FALSE) +
          scale_x_discrete(drop=FALSE)+
  ggtitle("Wilcoxon: Adjusted p-value distribution (Male control vs. Male 8 wk)")+
  theme(plot.title = element_text(hjust = 0.5))+
  labs(x="Adjusted p-value bins", y="Number peaks")


#############

# FEMALE
# create subsets of IDs 
controlids <- rownames(female[female$group=="control",])
week1_ids <- rownames(female[female$group=="1w",])
week2_ids <- rownames(female[female$group=="2w",])
week4_ids <- rownames(female[female$group=="4w",])
week8_ids <- rownames(female[female$group=="8w",])

# Note: filtered_norm_counts stores normalized count data for promoter regions only; could also try this with ATAC_KIDNEY_NORM_DATA (normalized data for all peaks, even omoter)
# Run Wilcoxon rank sum test (control vs. 8 wk):

todf <- list()
# for loop through all genes
allgenes <- rownames(F_filtered_norm_counts)
for (i in seq(length(allgenes))){
  
  # calculate log2fc between control and 4 wk
  gene <- allgenes[i]
  control <- as.numeric(F_filtered_norm_counts[i,controlids])
  wk8 <- as.numeric(F_filtered_norm_counts[i,week8_ids])
  log2fc <- mean(control) - mean(wk8)   #log(control/wk8) = log(control)-log(4wk)
  
  # wilcoxon rank sums - outputs tstatistic and pvalue; extract pvalue
  res <- wilcox.test(control, wk8, alternative = "two.sided", exact=FALSE)
  pvalue <- res$p.value
  
  # store results
  todf[[i]] <- c(gene, log2fc, pvalue)
}

# reformat list of vectors into dataframe
dfstats <- as.data.frame(do.call(rbind, todf))
colnames(dfstats) <- c("gene", "log2fc", "pvalue")
dfstats$log2fc <- as.numeric(dfstats$log2fc)
dfstats$pvalue <- as.numeric(dfstats$pvalue)
dfstats <- dfstats[order(dfstats$pvalue),]

# multiple hypothesis correction using benjamini-hochberg procedure
dfstats$qvalue <- p.adjust(dfstats$pvalue, method="BH")
head(dfstats)
dfstats[(dfstats$qvalue<0.05),]
any(dfstats$qvalue < 0.05) # NO SIGNIFICANT P-VALUES AFTER ADJUSTMENT 


# VISUALIZATION OF P-VALUES:
# Make control_8w contrast results into dataframe
dfstats$pbins<-NULL
dfstats$pbins[dfstats$qvalue > 0 & dfstats$qvalue < 0.05] <- "<0.05"
dfstats$pbins[dfstats$qvalue >=0.05 & dfstats$qvalue < 0.3] <- "0.05-0.3"
dfstats$pbins[dfstats$qvalue >=0.30 & dfstats$qvalue < 0.6] <- "0.3-0.6"
dfstats$pbins[dfstats$qvalue >=0.60 & dfstats$qvalue < 0.9] <- "0.6-.9"
dfstats$pbins[dfstats$qvalue >=0.90 & dfstats$qvalue <= 1] <- "0.9-1.0"

dfstats<-dfstats %>% 
  filter(is.na(qvalue) == FALSE)

ggplot(data = dfstats, 
       aes(x = pbins)) + 
  geom_histogram(stat = "count", color="lightpink", fill="lightpink") +
  stat_count(binwidth = 1, 
             geom = 'text', 
             color = 'black', 
             aes(label = ..count..),
           position = position_stack(vjust = 0.5))+
  scale_fill_discrete(drop=FALSE) +
          scale_x_discrete(drop=FALSE)+
  ggtitle("Wilcoxon: Adjusted p-value distribution (Female control vs. Female 8 wk)")+
  theme(plot.title = element_text(hjust = 0.5))+
  labs(x="Adjusted p-value bins", y="Number peaks")


```


Now we try grouping peaks by their nearest gene
```{r}
feature_ensembl <- ATAC_FEATURE_ANNOT %>% dplyr::select("feature_ID","ensembl_gene")

# Join the peaks with the ensembl annotation
combined_data_gene <- left_join(feature_ensembl, ATAC_KIDNEY_RAW_COUNTS) %>% dplyr:: select(-c("feature","assay","tissue"))
rownames(combined_data_gene) <- combined_data_gene$feature_ID
combined_data_gene <- combined_data_gene %>% dplyr::select(-c("feature_ID"))
dim(combined_data_gene)
# Now group by gene
grouped <- aggregate(. ~ ensembl_gene, combined_data_gene, sum)
rownames(grouped) <- grouped$ensembl_gene
grouped <- grouped %>% dplyr::select(-c("ensembl_gene"))

head(grouped)

```

Run DESeq on the grouped data
```{r de}
# Run DESeq2 on filtered raw counts and phenodata
dds <- DESeqDataSetFromMatrix(countData = grouped, 
                              colData = kidney_PHENO2,
                              design = ~Sample_batch+peak_enrich.frac_reads_in_peaks.macs2.frip+sex+sex*group+group)

dds <- DESeq(dds)

res <- results(dds)

summary(res)

# Look at contrasts between groups

# Control vs. 1 wk: results and top 10 by adjusted p-value

control_1w <-results(dds, contrast=c("group","control","1w"), tidy = TRUE) 
summary(control_1w)
head(control_1w[order(control_1w$padj),], 10)


# Control vs. 2 week
control_2w <- results(dds, contrast=c("group", "control", "2w"), tidy = TRUE)
summary(control_2w)
head(control_2w[order(control_2w$padj),], 10)

# Control vs. 4 week
control_4w <- results(dds, contrast=c("group", "control", "4w"), tidy = TRUE)
summary(control_4w)
head(control_4w[order(control_4w$padj),], 10)

# Control vs. 8 week
control_8w <- results(dds, contrast=c("group", "control", "8w"),tidy = TRUE)
summary(control_8w)
head(control_8w[order(control_8w$padj),], 10)

#Volcano for Control vs. 8 week
control_8w <- control_8w[order(res$padj),]
rownames(control_8w) <- control_8w$row
head(control_8w)

EnhancedVolcano(control_8w,
    lab = rownames(control_8w),
    x = 'log2FoldChange',
    y = 'padj')

# VISUALIZATION OF P-VALUES:
# Make control_8w contrast results into dataframe
control_8w<-data.frame(control_8w)
control_8w$pbins<-NULL
control_8w$pbins[control_8w$padj > 0 & control_8w$padj < 0.05] <- "<0.05"
control_8w$pbins[control_8w$padj >=0.05 & control_8w$padj < 0.3] <- "0.05-0.3"
control_8w$pbins[control_8w$padj >=0.30 & control_8w$padj < 0.6] <- "0.3-0.6"
control_8w$pbins[control_8w$padj >=0.60 & control_8w$padj < 0.9] <- "0.6-.9"
control_8w$pbins[control_8w$padj >=0.90 & control_8w$padj < 1] <- "0.9-1.0"

control_8w<-control_8w %>% 
  filter(is.na(padj) == FALSE)

ggplot(data = control_8w, 
       aes(x = pbins)) + 
  geom_histogram(stat = "count", color="light gray", fill="light gray") +
  stat_count(binwidth = 1, 
             geom = 'text', 
             color = 'black', 
             aes(label = ..count..),
           position = position_stack(vjust = 0.5))+
  scale_fill_discrete(drop=FALSE) +
          scale_x_discrete(drop=FALSE)+
  ggtitle("DESeq2: Adjusted p-value distribution (Control vs. 8 wk)")+
  theme(plot.title = element_text(hjust = 0.5))+
  labs(x="Adjusted p-value bins", y="Number peaks")




```

Try grouped data in limma.
```{r}

# Convert to factor
kidney_PHENO2$Sample_batch<-as.factor(kidney_PHENO2$Sample_batch)

# Create design matrix with model.matrix() -- specifies which variables/coefficients will be in linear model
full_model_str = "~ 1 + sex+ group + sex*group + Sample_batch + peak_enrich.frac_reads_in_peaks.macs2.frip"
design = model.matrix(as.formula(full_model_str), data=kidney_PHENO2)
design

# Limma model
v = voom(grouped, design, normalize.method="quantile")
limma_model1 = lmFit(v, design)

# contrasts from limma_fit1: "group2w"   "group4w"    "group8w"  "groupcontrol"  "sexmale:group2w"  "sexmale:group4w" "sexmale:group8w" "sexmale:groupcontrol"
limma_fit1 = contrasts.fit(limma_model1, coefficients=colnames(design)[grepl('group',colnames(design))]) 

eb_Ftest1 = eBayes(limma_fit1)
res = topTable(eb_Ftest1, n=nrow(eb_Ftest1))

# Highest log fold changes:
top.table<-topTreat(eb_Ftest1, sort.by = "logFC", n=Inf)
head(top.table)
length(which(top.table$adj.P.Val < 0.05)) # number of DEGs

```

Now we do GSEA! First, we convert the rat genes to human analogs so we can use human annotated gene sets

Our code follows this vignette https://stephenturner.github.io/deseq-to-fgsea/
```{r}
library(biomaRt)
mart <- useDataset("rnorvegicus_gene_ensembl", mart=useMart("ensembl"))
bm <- getBM(attributes=c("ensembl_gene_id", "hsapiens_homolog_associated_gene_name"), mart=mart) %>%
  distinct() %>%
  as_tibble()
bm

```

Then join the DESeq result with the gene symbols
```{r}
control_8w
res <- inner_join(control_8w, bm, by=c("row"="ensembl_gene_id"))
res2 <- res %>% 
  dplyr::select(hsapiens_homolog_associated_gene_name, stat) %>%
  distinct() %>% 
  group_by(hsapiens_homolog_associated_gene_name) %>% 
  summarize(stat=mean(stat)) %>% 
  filter(hsapiens_homolog_associated_gene_name != "")

```

Then use fgsea. First we checked out the hallmark 50 genesets
```{r}
library(fgsea)
ranks <- deframe(res2)
pathways.hallmark <- gmtPathways("h.all.v2023.1.Hs.symbols.gmt.txt")
fgseaRes <- fgsea(pathways=pathways.hallmark, stats=ranks)

fgsea_results <- as.data.frame(fgseaRes) %>% arrange(padj)
head(fgsea_results, 10)
```

```{r}
fgsea_results_sorted <- fgsea_results %>% arrange(NES)
ggplot(fgsea_results_sorted, aes(x=NES,y=pathway)) +
  scale_y_discrete(limits=fgsea_results_sorted$pathway) +
  geom_col(aes(fill=padj<0.05)) +
  theme(text = element_text(size = 6)) 
```

And we also looked at gene ontology annotated genesets
```{r}
library(fgsea)
ranks <- deframe(res2)
pathways.hallmark2 <- gmtPathways("c5.all.v2023.1.Hs.symbols.gmt.txt")
fgseaRes2 <- fgsea(pathways=pathways.hallmark2, stats=ranks)

fgsea_results2 <- as.data.frame(fgseaRes2) %>% arrange(padj)
head(fgsea_results2, 10)
```

All the above was for 8w, now check out the other time points too.
```{r}
# Control vs 1 week
res <- inner_join(control_1w, bm, by=c("row"="ensembl_gene_id"))
res2 <- res %>% 
  dplyr::select(hsapiens_homolog_associated_gene_name, stat) %>%
  distinct() %>% 
  group_by(hsapiens_homolog_associated_gene_name) %>% 
  summarize(stat=mean(stat)) %>% 
  filter(hsapiens_homolog_associated_gene_name != "")

ranks <- deframe(res2)
pathways.hallmark <- gmtPathways("h.all.v2023.1.Hs.symbols.gmt.txt")
fgseaRes <- fgsea(pathways=pathways.hallmark, stats=ranks)
fgsea_results <- as.data.frame(fgseaRes) %>% arrange(padj)
head(fgsea_results, 10)


fgsea_results_sorted <- fgsea_results %>% arrange(NES)
ggplot(fgsea_results_sorted, aes(x=NES,y=pathway)) +
  scale_y_discrete(limits=fgsea_results_sorted$pathway) +
  geom_col(aes(fill=padj<0.05)) +
  theme(text = element_text(size = 6)) 
```

```{r}
# Control vs 2 week
res <- inner_join(control_2w, bm, by=c("row"="ensembl_gene_id"))
res2 <- res %>% 
  dplyr::select(hsapiens_homolog_associated_gene_name, stat) %>%
  distinct() %>% 
  group_by(hsapiens_homolog_associated_gene_name) %>% 
  summarize(stat=mean(stat)) %>% 
  filter(hsapiens_homolog_associated_gene_name != "")

ranks <- deframe(res2)
pathways.hallmark <- gmtPathways("h.all.v2023.1.Hs.symbols.gmt.txt")
fgseaRes <- fgsea(pathways=pathways.hallmark, stats=ranks)
fgsea_results <- as.data.frame(fgseaRes) %>% arrange(padj)
head(fgsea_results, 10)


fgsea_results_sorted <- fgsea_results %>% arrange(NES)
ggplot(fgsea_results_sorted, aes(x=NES,y=pathway)) +
  scale_y_discrete(limits=fgsea_results_sorted$pathway) +
  geom_col(aes(fill=padj<0.05)) +
  theme(text = element_text(size = 6)) 
```

```{r}
# Control vs 4 week
res <- inner_join(control_4w, bm, by=c("row"="ensembl_gene_id"))
res2 <- res %>% 
  dplyr::select(hsapiens_homolog_associated_gene_name, stat) %>%
  distinct() %>% 
  group_by(hsapiens_homolog_associated_gene_name) %>% 
  summarize(stat=mean(stat)) %>% 
  filter(hsapiens_homolog_associated_gene_name != "")

ranks <- deframe(res2)
pathways.hallmark <- gmtPathways("h.all.v2023.1.Hs.symbols.gmt.txt")
fgseaRes <- fgsea(pathways=pathways.hallmark, stats=ranks)
fgsea_results <- as.data.frame(fgseaRes) %>% arrange(padj)
head(fgsea_results, 10)


fgsea_results_sorted <- fgsea_results %>% arrange(NES)
ggplot(fgsea_results_sorted, aes(x=NES,y=pathway)) +
  scale_y_discrete(limits=fgsea_results_sorted$pathway) +
  geom_col(aes(fill=padj<0.05)) +
  theme(text = element_text(size = 6)) 
```


































